---
title: "MDV Nutrient Limitation Rank of ASVs by Basin-Treatment"
author: "Adam Solon"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---

<h1 align = "center" > Top 10 ASVs by relative abundance (%) </h1>
&nbsp;  
&nbsp;    

### Script Summary  
This script determines the top 10 Ampicon Sequence Variants (ASVs) for each level of a factor (e.g., treatments in an experiment, habitats in a local area). To run this script you will need the output data frame from DADA2 that includes ASV#, sequence counts, and taxonomic assignment.

##### Ten steps to running this pipeline:  
1.  Organizing and formatting files
2.  Subsetting according to basin


```{r, echo = FALSE, include = FALSE}
# Change identifiers to your system and file naming. 
user <- "C:/Users/adams" # computer user name
folder <- "/Desktop/NAM"
project <- "/manuscript1"
analyses <- "/analyses"
data  <- "/dna_seq/amplicons"
analysis2 <- "/RA_relative_abundance/16S"
ecosystem <- "high elevation soils" # Define the environment for use in report language.
project.fp <- paste0(user, folder, project)
```

```{r Install and load packages, echo = FALSE, include = FALSE}

# install.packages("tidyverse")
# install.packages("knitr")
# install.packages("kableExtra")
# install.packages("plotly")
# install.packages("stringr")

library(tidyverse)
library(knitr)
library(kableExtra)
library(plotly)
library(stringr)

knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(include = FALSE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)
```

### Step 1: Organize project directory, input files, and format

The first step to running this pipeline is to organize folders and identify file paths. This keeps everything organized and helps further down the pipeline in retrieving and saving results in a reliable location. 

```{r set paths for project}
# First define the project and project folders. 

# Create project directory
###################################################
project.fp <- paste0(user, folder, project)
if (!dir.exists(project.fp)) dir.create(project.fp)

# Create sub-directory for all analyses for project 
###################################################
analyses.fp <- paste0(project.fp, analyses)
if (!dir.exists(analyses.fp)) dir.create(analyses.fp)

# Create sub-directory for data type 
###################################################
data.fp <- paste0(analyses.fp, data)
if (!dir.exists(data.fp)) dir.create(data.fp)

# Create sub-directory for specific analysis 
###################################################
analysis2.fp <- paste0(data.fp, analysis2)
if (!dir.exists(analysis2.fp)) dir.create(analysis2.fp)

# Create sub-directories for analysis 
###################################################

# Create directory for 'before' pipeline inputs
input.fp <- paste0(analysis2.fp, "/input")
if (!dir.exists(input.fp)) dir.create(input.fp)

# Create directory for 'within' pipeline R objects 
objects.fp <- paste0(analysis2.fp, "/objects")
if (!dir.exists(objects.fp)) dir.create(objects.fp)

# Create directory for 'after' pipeline outputs 
output.fp <- paste0(analysis2.fp, "/output")
if (!dir.exists(output.fp)) dir.create(output.fp)

# Check for the folders here as a sanity check. Should see "Input" and "Objects" if starting from scratch.
list.files(analysis2.fp) 
```

```{r set paths for objects and output by basin}
# Set up output sub-directories
# Create sub-directories within the objects directory
################################################### 
fryxell1.fp <- paste0(objects.fp, "/fryxell")
if (!dir.exists(fryxell1.fp)) dir.create(fryxell1.fp)

figures.fp <- paste0(objects.fp, "/figures") 
if (!dir.exists(figures.fp)) dir.create(figures.fp)

# Check for the Output folder 
list.files(objects.fp) 

# Create sub-directories within the output directory
################################################### 
fryxell2.fp <- paste0(output.fp, "/fryxell")
if (!dir.exists(fryxell2.fp)) dir.create(fryxell2.fp)

figures.fp <- paste0(output.fp, "/figures") 
if (!dir.exists(figures.fp)) dir.create(figures.fp)

# Check for the Output folder 
list.files(output.fp) 
```

#### Input files   
Required input files:   
  
1.  The ASV table from DADA2 pipeline 
2.  The 'mapping file' w/ relevant metadata for each sample

Input files and re-format for use in pipeline

```{r set paths to raw data and create input objects}
# ASV Table (from DADA2)
asvTable.BAC.fp <- paste0(input.fp, "/ASVtable_16S_wTax_fryxell.txt") 

# Mapping File (metadata relevant for study samples)
mappingFile.fp <- paste0(input.fp, "/map_file_fryxell.txt") 

#input 16s ASV table w/ taxonomy
asvTab.BAC <- read.table(asvTable.BAC.fp,  header = T, sep = "\t")

#input metadata
mapfile <- read.table(mappingFile.fp, header = T, sep = "\t")

```

#### Format files
```{r create new data frame for taxonomy seprate from ASV count table}
# remove taxonomy from ASV table 
taxTab <- asvTab.BAC %>% select(Domain:Genus)

# create data.frame w/ ASV ID as row names
rownames(taxTab) <- asvTab.BAC$ASV_ID

# move row names (ASV ID) to 1st column
taxTab <- tibble::rownames_to_column(taxTab, "ASV_ID")

#remove taxonomy column from ASV table
asvTab.BAC.1 <- asvTab.BAC %>% select(!Domain:Genus)

#set ASV IDs as row names
rownames(asvTab.BAC.1) <- asvTab.BAC.1$ASV_ID

#remove ASV IDs column
asvTab.BAC.1$ASV_ID <- NULL

#transpose so columns (SampleIDs) are rows
asvTab.BAC.2 <- as.data.frame(t(asvTab.BAC.1))

#move row names into 1st column
asvTab.BAC.2 <- asvTab.BAC.2 %>% rownames_to_column("SampleID")
```

####Subset samples
1. Subset samples from basin into experimental groups (i.e., Bulk, CTRL, N, P, NP)

```{r subset mapping file and ASV tables according to basin}
# subset mapping file by treatment

# subset 'bulk'
#####################################################
fryxell.b <- filter(mapfile, treatment == "bulk") 

# create object w/ dplyr to subset ASV table
samp.sub.b.b <- dplyr::pull(fryxell.b, SampleID)

# subset samples by using only samples IDs in chosen mapping file
asvTab.BAC.b.b <- asvTab.BAC.2[asvTab.BAC.2$SampleID %in% samp.sub.b.b, ]

# move sample IDs back to row names 
rownames(asvTab.BAC.b.b) <- asvTab.BAC.b.b$SampleID

# remove sample ID column
asvTab.BAC.b.b$SampleID <- NULL

# transpose ASV table back to samples as columns (ASVs as rows)
asvTab.BAC.b.b.1 <- as.data.frame(t(asvTab.BAC.b.b))

# subset 'controls' (i.e., microcosms w/ no nutrients added)
#####################################################
fryxell.c <- filter(mapfile, treatment == "CTRL") 

# create object w/ dplyr to subset ASV table
samp.sub.b.c <- dplyr::pull(fryxell.c, SampleID)

# subset samples by using only samples IDs in chosen mapping file
asvTab.BAC.b.c <- asvTab.BAC.2[asvTab.BAC.2$SampleID %in% samp.sub.b.c, ]

# move sample IDs back to row names 
rownames(asvTab.BAC.b.c) <- asvTab.BAC.b.c$SampleID

# remove sample ID column
asvTab.BAC.b.c$SampleID <- NULL

# transpose ASV table back to samples as columns (ASVs as rows)
asvTab.BAC.b.c.1 <- as.data.frame(t(asvTab.BAC.b.c))

#subset N treatment (i.e., treatments with 30ug/g N)
#####################################################
fryxell.n <- filter(mapfile, treatment == "N") 

# create object w/ dplyr to subset ASV table
samp.sub.b.n <- dplyr::pull(fryxell.n, SampleID)

# subset samples by using only samples IDs in chosen mapping file
asvTab.BAC.b.n <- asvTab.BAC.2[asvTab.BAC.2$SampleID %in% samp.sub.b.n, ]

# move sample IDs back to row names 
rownames(asvTab.BAC.b.n) <- asvTab.BAC.b.n$SampleID

# remove sample ID column
asvTab.BAC.b.n$SampleID <- NULL

# transpose ASV table back to samples as columns (ASVs as rows)
asvTab.BAC.b.n.1 <- as.data.frame(t(asvTab.BAC.b.n))

#subset P treatment (i.e., treatments with 30ug/g P)
#####################################################
fryxell.p <- filter(mapfile, treatment == "P") 

# create object w/ dplyr to subset ASV table
samp.sub.b.p <- dplyr::pull(fryxell.p, SampleID)

# subset samples by using only samples IDs in chosen mapping file
asvTab.BAC.b.p <- asvTab.BAC.2[asvTab.BAC.2$SampleID %in% samp.sub.b.p, ]

# move sample IDs back to row names 
rownames(asvTab.BAC.b.p) <- asvTab.BAC.b.p$SampleID

# remove sample ID column
asvTab.BAC.b.p$SampleID <- NULL

# transpose ASV table back to samples as columns (ASVs as rows)
asvTab.BAC.b.p.1 <- as.data.frame(t(asvTab.BAC.b.p))

#subset N & P treatment (i.e., treatments with 30ug/g N + 30ug/g P)
#####################################################
fryxell.np <- filter(mapfile, treatment == "NP") 

# create object w/ dplyr to subset ASV table
samp.sub.b.np <- dplyr::pull(fryxell.np, SampleID)

# subset samples by using only samples IDs in chosen mapping file
asvTab.BAC.b.np <- asvTab.BAC.2[asvTab.BAC.2$SampleID %in% samp.sub.b.np, ]

# move sample IDs back to row names 
rownames(asvTab.BAC.b.np) <- asvTab.BAC.b.np$SampleID

# remove sample ID column
asvTab.BAC.b.np$SampleID <- NULL

# transpose ASV table back to samples as columns (ASVs as rows)
asvTab.BAC.b.np.1 <- as.data.frame(t(asvTab.BAC.b.np))


```

```{r create new columns with row sums and reorder rows by descending value}
# Bulk samples
########################################################
# create row sum column
asvTab.BAC.b.b.2 <- asvTab.BAC.b.b.1 %>% mutate(total = rowSums(.))

# re-order with greatest row sum 1st
asvTab.BAC.b.b.2 <- asvTab.BAC.b.b.2 %>% arrange(desc(total))

# CTRL samples
########################################################
# create row sum column
asvTab.BAC.b.c.2 <- asvTab.BAC.b.c.1 %>% mutate(total = rowSums(.))

# re-order with greatest row sum 1st
asvTab.BAC.b.c.2 <- asvTab.BAC.b.c.2 %>% arrange(desc(total))

# N samples
########################################################
# create row sum column
asvTab.BAC.b.n.2 <- asvTab.BAC.b.n.1 %>% mutate(total = rowSums(.))

# re-order with greatest row sum 1st
asvTab.BAC.b.n.2 <- asvTab.BAC.b.n.2 %>% arrange(desc(total))

# P samples
########################################################
# create row sum column
asvTab.BAC.b.p.2 <- asvTab.BAC.b.p.1 %>% mutate(total = rowSums(.))

# re-order with greatest row sum 1st
asvTab.BAC.b.p.2 <- asvTab.BAC.b.p.2 %>% arrange(desc(total))

# NP samples
########################################################
# create row sum column
asvTab.BAC.b.np.2 <- asvTab.BAC.b.np.1 %>% mutate(total = rowSums(.))

# re-order with greatest row sum 1st
asvTab.BAC.b.np.2 <- asvTab.BAC.b.np.2 %>% arrange(desc(total))

```

```{r remove rows (i.e., ASVs) that contain zeros}
# Bulk samples
########################################################
# create row sum column
asvTab.BAC.b.b.3 <- asvTab.BAC.b.b.2 %>% filter(total!= 0)

# CTRL samples
########################################################
# create row sum column
asvTab.BAC.b.c.3 <- asvTab.BAC.b.c.2 %>% filter(total!= 0)

# N samples
########################################################
# create row sum column
asvTab.BAC.b.n.3 <- asvTab.BAC.b.n.2 %>% filter(total!= 0)

# P samples
########################################################
# create row sum column
asvTab.BAC.b.p.3 <- asvTab.BAC.b.p.2 %>% filter(total!= 0)

# NP samples
########################################################
# create row sum column
asvTab.BAC.b.np.3 <- asvTab.BAC.b.np.2 %>% filter(total!= 0)

```

```{r calculate relative abundance (%)}
# Bulk samples
########################################################
# calculate relative abundance of each cell by column value
asvTab.BAC.b.b.4 <- asvTab.BAC.b.b.3 %>% apply(2, function(x){x/sum(x)}) 
# multiply by 100 for percentage
asvTab.BAC.b.b.5 <- asvTab.BAC.b.b.4 %>% apply(2, function(x){x *100})
# classify as a data frame
asvTab.BAC.b.b.5 <- as.data.frame(asvTab.BAC.b.b.5)

# CTRL samples
########################################################
# calculate relative abundance of each cell by column value
asvTab.BAC.b.c.4 <- asvTab.BAC.b.c.3 %>% apply(2, function(x){x/sum(x)})
# multiply by 100 for percentage
asvTab.BAC.b.c.5 <- asvTab.BAC.b.c.4 %>% apply(2, function(x){x *100})
# classify as a data frame
asvTab.BAC.b.c.5 <- as.data.frame(asvTab.BAC.b.c.5)

# N samples
########################################################
# calculate relative abundance of each cell by column value
asvTab.BAC.b.n.4 <- asvTab.BAC.b.n.3 %>% apply(2, function(x){x/sum(x)})
# multiply by 100 for percentage
asvTab.BAC.b.n.5 <- asvTab.BAC.b.n.4 %>% apply(2, function(x){x *100})
# classify as a data frame
asvTab.BAC.b.n.5 <- as.data.frame(asvTab.BAC.b.n.5)

# P samples
########################################################
# calculate relative abundance of each cell by column value
asvTab.BAC.b.p.4 <- asvTab.BAC.b.p.3 %>% apply(2, function(x){x/sum(x)})
# multiply by 100 for percentage
asvTab.BAC.b.p.5 <- asvTab.BAC.b.p.4 %>% apply(2, function(x){x *100})
# classify as a data frame
asvTab.BAC.b.p.5 <- as.data.frame(asvTab.BAC.b.p.5)

# NP samples
########################################################
# calculate relative abundance of each cell by column value
asvTab.BAC.b.np.4 <- asvTab.BAC.b.np.3 %>% apply(2, function(x){x/sum(x)})
# multiply by 100 for percentage
asvTab.BAC.b.np.5 <- asvTab.BAC.b.np.4 %>% apply(2, function(x){x *100})
# classify as a data frame
asvTab.BAC.b.np.5 <- as.data.frame(asvTab.BAC.b.np.5)

```

```{r remove some samples (columns) that create XXX issues}
# Bulk samples extracted with different extraction kit
asvTab.BAC.b.b.F <- asvTab.BAC.b.b.1 %>% select(-c(FX_Bulk_X, FX_Bulk_Y, FX_Bulk_Z))

```

```{r create new columns with row sums and reorder rows by descending value}
# Bulk samples
########################################################
# create row sum column
asvTab.BAC.b.b.F.1 <- asvTab.BAC.b.b.F %>% mutate(total = rowSums(.))

# re-order with greatest row sum 1st
asvTab.BAC.b.b.F.2 <- asvTab.BAC.b.b.F.1 %>% arrange(desc(total))

```


```{r remove rows (i.e., ASVs) that contain zeros}
# Bulk samples
########################################################
# create row sum column
asvTab.BAC.b.b.F.3 <- asvTab.BAC.b.b.F.2 %>% filter(total!= 0)

```


```{r calculate relative abundance (%)}
# Bulk samples
########################################################
# calculate relative abundance of each cell by column value
asvTab.BAC.b.b.F.4 <- asvTab.BAC.b.b.F.3 %>% apply(2, function(x){x/sum(x)}) 
# multiply by 100 for percentage
asvTab.BAC.b.b.F.5 <- asvTab.BAC.b.b.F.4 %>% apply(2, function(x){x *100})
# classify as a data frame
asvTab.BAC.b.b.F.5 <- as.data.frame(asvTab.BAC.b.b.F.5)

```

####Add Taxonomy back-in to identify ESVs

```{r merge the taxonomy data frame w/ the ESV relative abundance (%) data frames}
# Bulk samples
########################################################
# move rownames of ASV table into 1st column as 'ASV_ID'
asvTab.BAC.b.b.F.5 <- tibble::rownames_to_column(asvTab.BAC.b.b.F.5, "ASV_ID")
# Left Join using inner_join function from dplyr
asvTab.BAC.b.b.F.T <- asvTab.BAC.b.b.F.5 %>% inner_join(taxTab, by= "ASV_ID")

# CTRL samples
########################################################
# move rownames of ASV table into 1st column as 'ASV_ID'
asvTab.BAC.b.c.5 <- tibble::rownames_to_column(asvTab.BAC.b.c.5, "ASV_ID")
# Left Join using inner_join function from dplyr
asvTab.BAC.b.c.5.T <- asvTab.BAC.b.c.5 %>% inner_join(taxTab, by= "ASV_ID")

# N samples
########################################################
# move rownames of ASV table into 1st column as 'ASV_ID'
asvTab.BAC.b.n.5 <- tibble::rownames_to_column(asvTab.BAC.b.n.5, "ASV_ID")
# Left Join using inner_join function from dplyr
asvTab.BAC.b.n.5.T <- asvTab.BAC.b.n.5 %>% inner_join(taxTab, by= "ASV_ID")

# P samples
########################################################
# move rownames of ASV table into 1st column as 'ASV_ID'
asvTab.BAC.b.p.5 <- tibble::rownames_to_column(asvTab.BAC.b.p.5, "ASV_ID")
# Left Join using inner_join function from dplyr
asvTab.BAC.b.p.5.T <- asvTab.BAC.b.p.5 %>% inner_join(taxTab, by= "ASV_ID")

# NP samples
########################################################
# move rownames of ASV table into 1st column as 'ASV_ID'
asvTab.BAC.b.np.5 <- tibble::rownames_to_column(asvTab.BAC.b.np.5, "ASV_ID")
# Left Join using inner_join function from dplyr
asvTab.BAC.b.np.F.T <- asvTab.BAC.b.np.5 %>% inner_join(taxTab, by= "ASV_ID")

```

Save as R objects and .txt files

```{r}
# Bulk samples
########################################################
# save as an R file
saveRDS(asvTab.BAC.b.b.F.T, paste0(fryxell1.fp, "/asvTab.BAC.b.b.F.T.rds"))
#save as .txt file
write.table(asvTab.BAC.b.b.F.T, file = paste0(fryxell2.fp, "/FryxellBulk_RelativeAbundance.txt"), 
            sep = "\t", row.names = TRUE, col.names = NA)

# CTRL samples
########################################################
# save as an R file
saveRDS(asvTab.BAC.b.c.5.T, paste0(fryxell1.fp, "/asvTab.BAC.b.c.5.T.rds"))
#save as .txt file
write.table(asvTab.BAC.b.c.5.T, file = paste0(fryxell2.fp, "/FryxellCTRL_RelativeAbundance.txt"), 
            sep = "\t", row.names = TRUE, col.names = NA)

# N samples
########################################################
# save as an R file
saveRDS(asvTab.BAC.b.n.5.T, paste0(fryxell1.fp, "/asvTab.BAC.b.n.5.T.rds"))
#save as .txt file
write.table(asvTab.BAC.b.n.5.T, file = paste0(fryxell2.fp, "/FryxellN_RelativeAbundance.txt"), 
            sep = "\t", row.names = TRUE, col.names = NA)

# P samples
########################################################
# save as an R file
saveRDS(asvTab.BAC.b.p.5.T, paste0(fryxell1.fp, "/asvTab.BAC.b.p.5.T.rds"))
#save as .txt file
write.table(asvTab.BAC.b.p.5.T, file = paste0(fryxell2.fp, "/FryxellP_RelativeAbundance.txt"), 
            sep = "\t", row.names = TRUE, col.names = NA)

# NP samples
########################################################
# save as an R file
saveRDS(asvTab.BAC.b.np.F.T, paste0(fryxell1.fp, "/asvTab.BAC.b.np.F.T.rds"))
#save as .txt file
write.table(asvTab.BAC.b.np.F.T, file = paste0(fryxell2.fp, "/FryxellNP_RelativeAbundance.txt"), 
            sep = "\t", row.names = TRUE, col.names = NA)
```