---
title: "Nutrient Limitation in the MDV of Antarctica"
subtitle: "  \n Top ASVs Hoare Basin - 16S SSU rRNA gene"
author: "Adam J. Solon"
date: "`r Sys.Date()`"
#output: html_document
output: 
  pdf_document:
    #includes:  
     # in_header: my_header.txt
    toc: TRUE
    fig_width: 8
    fig_height: 6
    fig_caption: true
fontsize: 12pt
#editor_options: 
#  chunk_output_type: console
---

# Script Summary  
This script processes the Amplicon Sequence Variants (ASVs) table from a nutrient limitation study of soils of Hoare basin from Taylor Valley, antarctica.  It will determine mean relative abundances of ASVs in each experimental group (i.e., bulk, control, nitrogen, phosphorus, and nitrogen&phosphorus).

### Steps of this pipeline:  
1.  Create and organize directories
2.  Load R packages
3.  Input files
4.  Format Files
5.  Subset according to group
6.  Calculate relative abundance
7.  Save files
  
```{r echo = FALSE, include = FALSE}
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(include = FALSE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)

```

```{r}
# Change identifiers to your system and file naming. 
user <- "C:/Users/adams" # computer user name
folder <- "/Desktop/Projects"
project <- "/MDV_Soils_nutrient_limitation/manuscript1"
analyses <- "/analyses"
data  <- "/dna_seq/amplicons"
analysis2 <- "/RA_relative_abundance/16S"
ecosystem <- "lab microcosms - Taylor Valley soils" # Define the environment for use in report language.
project.fp <- paste0(user, folder, project)

```
  
```{r set paths for project}
# First define the project and project folders. 

# Create project directory
###################################################
project.fp <- paste0(user, folder, project)
if (!dir.exists(project.fp)) dir.create(project.fp)

# Create sub-directory for all analyses for project 
###################################################
analyses.fp <- paste0(project.fp, analyses)
if (!dir.exists(analyses.fp)) dir.create(analyses.fp)

# Create sub-directory for data type 
###################################################
data.fp <- paste0(analyses.fp, data)
if (!dir.exists(data.fp)) dir.create(data.fp)

# Create sub-directory for Relative Abundance (RA) analysis
###################################################
analysis2.fp <- paste0(data.fp, analysis2)
if (!dir.exists(analysis2.fp)) dir.create(analysis2.fp)

# Create sub-directories for RA 
###################################################

# Create directory for 'before' pipeline inputs
input2.fp <- paste0(analysis2.fp, "/input")
if (!dir.exists(input2.fp)) dir.create(input2.fp)

# Create directory for 'within' pipeline R objects 
objects2.fp <- paste0(analysis2.fp, "/objects")
if (!dir.exists(objects2.fp)) dir.create(objects2.fp)

# Create directory for 'after' pipeline outputs 
output2.fp <- paste0(analysis2.fp, "/output")
if (!dir.exists(output2.fp)) dir.create(output2.fp)

# Check for the folders here as a sanity check. Should see "Input" and "Objects" if starting from scratch.
list.files(analysis2.fp) 

# Create further sub-directories for Hoare basin
###################################################

# Create directory for 'within' pipeline R objects 
objects2.h.fp <- paste0(objects2.fp, "/hoare")
if (!dir.exists(objects2.h.fp)) dir.create(objects2.h.fp)

# Create directory for 'after' pipeline outputs 
output2.h.fp <- paste0(output2.fp, "/hoare")
if (!dir.exists(output2.h.fp)) dir.create(output2.h.fp)

```
  
```{r Install and load packages}

# install.packages("tidyverse")
# install.packages("knitr")
# install.packages("kableExtra")
# install.packages("plotly")
# install.packages("stringr")

library(tidyverse); packageVersion("tidyverse")
library(knitr); packageVersion("knitr")
library(kableExtra); packageVersion("kableExtra")
library(stringr); packageVersion("stringr")

```
    
* r version: `r getRversion()`
* RStudio version: `r rstudioapi::versionInfo()$version`
* r packages:  
  tidyverse, `r packageVersion("tidyverse")`   
  knitr, `r packageVersion("knitr")`  
  kableExtra, `r packageVersion("KableExtra")`  
  stringr, `r packageVersion("stringr")`  
  
### Input Files
Required input files:   
  
1.  The ASV table from DADA2 pipeline 
2.  The 'mapping file' w/ relevant metadata for each sample

Input files and re-format for use in pipeline  
  
```{r set paths to raw data and create input objects}
# ASV Table (from DADA2)
asvTable.BAC.fp <- paste0(input2.fp, "/ASVtable_16S_wTax_hoare.txt") 

# Mapping File (metadata relevant for study samples)
mappingFile.fp <- paste0(input2.fp, "/map_file_hoare.txt") 

#input 16s ASV table w/ taxonomy
asvTab.BAC <- read.table(asvTable.BAC.fp,  header = T, sep = "\t")

#input metadata
mapfile <- read.table(mappingFile.fp, header = T, sep = "\t")

```
  
### Format Files  
  
```{r create new data frame for taxonomy seprate from ASV count table}
# remove taxonomy from ASV table 
taxTab <- asvTab.BAC %>% select(Domain:Genus)

# create data.frame w/ ASV ID as row names
rownames(taxTab) <- asvTab.BAC$ASV_ID

# move row names (ASV ID) to 1st column
taxTab <- tibble::rownames_to_column(taxTab, "ASV_ID")

#remove taxonomy column from ASV table
asvTab.BAC.1 <- asvTab.BAC %>% select(!Domain:Genus)

#set ASV IDs as row names
rownames(asvTab.BAC.1) <- asvTab.BAC.1$ASV_ID

#remove ASV IDs column
asvTab.BAC.1$ASV_ID <- NULL

#transpose so columns (SampleIDs) are rows
asvTab.BAC.2 <- as.data.frame(t(asvTab.BAC.1))

#move row names into 1st column
asvTab.BAC.2 <- asvTab.BAC.2 %>% rownames_to_column("SampleID")

```
  
### Subset by treatment  
Subset by experimental groups: bulk, control (H2O), nitrogen (N and H2O), phosphorus (P and H2O), and nitrogen&phosphorus (N, P, and H2O)  
  
```{r subset mapping file and ASV tables according to basin}
# subset mapping file by treatment

# subset 'bulk'
#####################################################
hoare.b <- filter(mapfile, treatment == "bulk") 

# create object w/ dplyr to subset ASV table
samp.sub.b.b <- dplyr::pull(hoare.b, SampleID)

# subset samples by using only samples IDs in chosen mapping file
asvTab.BAC.b.b <- asvTab.BAC.2[asvTab.BAC.2$SampleID %in% samp.sub.b.b, ]

# move sample IDs back to row names 
rownames(asvTab.BAC.b.b) <- asvTab.BAC.b.b$SampleID

# remove sample ID column
asvTab.BAC.b.b$SampleID <- NULL

# transpose ASV table back to samples as columns (ASVs as rows)
asvTab.BAC.b.b.1 <- as.data.frame(t(asvTab.BAC.b.b))

#recode values as numeric (from character)
asvTab.BAC.b.b.1 <- asvTab.BAC.b.b.1 %>% mutate_if(is.character,as.numeric)

# subset 'controls' (i.e., microcosms w/ no nutrients added)
#####################################################
hoare.c <- filter(mapfile, treatment == "CTRL") 

# create object w/ dplyr to subset ASV table
samp.sub.b.c <- dplyr::pull(hoare.c, SampleID)

# subset samples by using only samples IDs in chosen mapping file
asvTab.BAC.b.c <- asvTab.BAC.2[asvTab.BAC.2$SampleID %in% samp.sub.b.c, ]

# move sample IDs back to row names 
rownames(asvTab.BAC.b.c) <- asvTab.BAC.b.c$SampleID

# remove sample ID column
asvTab.BAC.b.c$SampleID <- NULL

# transpose ASV table back to samples as columns (ASVs as rows)
asvTab.BAC.b.c.1 <- as.data.frame(t(asvTab.BAC.b.c))

#recode values as numeric (from character)
asvTab.BAC.b.c.1 <- asvTab.BAC.b.c.1 %>% mutate_if(is.character,as.numeric)

#subset N treatment (i.e., treatments with 30ug/g N)
#####################################################
hoare.n <- filter(mapfile, treatment == "N") 

# create object w/ dplyr to subset ASV table
samp.sub.b.n <- dplyr::pull(hoare.n, SampleID)

# subset samples by using only samples IDs in chosen mapping file
asvTab.BAC.b.n <- asvTab.BAC.2[asvTab.BAC.2$SampleID %in% samp.sub.b.n, ]

# move sample IDs back to row names 
rownames(asvTab.BAC.b.n) <- asvTab.BAC.b.n$SampleID

# remove sample ID column
asvTab.BAC.b.n$SampleID <- NULL

# transpose ASV table back to samples as columns (ASVs as rows)
asvTab.BAC.b.n.1 <- as.data.frame(t(asvTab.BAC.b.n))

#recode values as numeric (from character)
asvTab.BAC.b.n.1 <- asvTab.BAC.b.n.1 %>% mutate_if(is.character,as.numeric)

#subset P treatment (i.e., treatments with 30ug/g P)
#####################################################
hoare.p <- filter(mapfile, treatment == "P") 

# create object w/ dplyr to subset ASV table
samp.sub.b.p <- dplyr::pull(hoare.p, SampleID)

# subset samples by using only samples IDs in chosen mapping file
asvTab.BAC.b.p <- asvTab.BAC.2[asvTab.BAC.2$SampleID %in% samp.sub.b.p, ]

# move sample IDs back to row names 
rownames(asvTab.BAC.b.p) <- asvTab.BAC.b.p$SampleID

# remove sample ID column
asvTab.BAC.b.p$SampleID <- NULL

# transpose ASV table back to samples as columns (ASVs as rows)
asvTab.BAC.b.p.1 <- as.data.frame(t(asvTab.BAC.b.p))

#recode values as numeric (from character)
asvTab.BAC.b.p.1 <- asvTab.BAC.b.p.1 %>% mutate_if(is.character,as.numeric)

#subset N & P treatment (i.e., treatments with 30ug/g N + 30ug/g P)
#####################################################
hoare.np <- filter(mapfile, treatment == "NP") 

# create object w/ dplyr to subset ASV table
samp.sub.b.np <- dplyr::pull(hoare.np, SampleID)

# subset samples by using only samples IDs in chosen mapping file
asvTab.BAC.b.np <- asvTab.BAC.2[asvTab.BAC.2$SampleID %in% samp.sub.b.np, ]

# move sample IDs back to row names 
rownames(asvTab.BAC.b.np) <- asvTab.BAC.b.np$SampleID

# remove sample ID column
asvTab.BAC.b.np$SampleID <- NULL

# transpose ASV table back to samples as columns (ASVs as rows)
asvTab.BAC.b.np.1 <- as.data.frame(t(asvTab.BAC.b.np))

#recode values as numeric (from character)
asvTab.BAC.b.np.1 <- asvTab.BAC.b.np.1 %>% mutate_if(is.character,as.numeric)

```
  
### Total ASVs across groups  
  
```{r create new columns with row sums and reorder rows by descending value}
# Bulk samples
########################################################
# create row sum column
asvTab.BAC.b.b.2 <- asvTab.BAC.b.b.1 %>% mutate(total = rowSums(.))

# re-order with greatest row sum 1st
asvTab.BAC.b.b.2 <- asvTab.BAC.b.b.2 %>% arrange(desc(total))

# CTRL samples
########################################################
# create row sum column
asvTab.BAC.b.c.2 <- asvTab.BAC.b.c.1 %>% mutate(total = rowSums(.))

# re-order with greatest row sum 1st
asvTab.BAC.b.c.2 <- asvTab.BAC.b.c.2 %>% arrange(desc(total))

# N samples
########################################################
# create row sum column
asvTab.BAC.b.n.2 <- asvTab.BAC.b.n.1 %>% mutate(total = rowSums(.))

# re-order with greatest row sum 1st
asvTab.BAC.b.n.2 <- asvTab.BAC.b.n.2 %>% arrange(desc(total))

# P samples
########################################################
# create row sum column
asvTab.BAC.b.p.2 <- asvTab.BAC.b.p.1 %>% mutate(total = rowSums(.))

# re-order with greatest row sum 1st
asvTab.BAC.b.p.2 <- asvTab.BAC.b.p.2 %>% arrange(desc(total))

# NP samples
########################################################
# create row sum column
asvTab.BAC.b.np.2 <- asvTab.BAC.b.np.1 %>% mutate(total = rowSums(.))

# re-order with greatest row sum 1st
asvTab.BAC.b.np.2 <- asvTab.BAC.b.np.2 %>% arrange(desc(total))

```

```{r remove rows (i.e., ASVs) that contain zeros}
# Bulk samples
########################################################
# create row sum column
asvTab.BAC.b.b.3 <- asvTab.BAC.b.b.2 %>% filter(total!= 0)

# CTRL samples
########################################################
# create row sum column
asvTab.BAC.b.c.3 <- asvTab.BAC.b.c.2 %>% filter(total!= 0)

# N samples
########################################################
# create row sum column
asvTab.BAC.b.n.3 <- asvTab.BAC.b.n.2 %>% filter(total!= 0)

# P samples
########################################################
# create row sum column
asvTab.BAC.b.p.3 <- asvTab.BAC.b.p.2 %>% filter(total!= 0)

# NP samples
########################################################
# create row sum column
asvTab.BAC.b.np.3 <- asvTab.BAC.b.np.2 %>% filter(total!= 0)

```
  
### Calculate ASV relative abundance per group  
  
```{r calculate relative abundance (%)}
# Bulk samples
########################################################
# calculate relative abundance of each cell by column value
asvTab.BAC.b.b.4 <- asvTab.BAC.b.b.3 %>% apply(2, function(x){x/sum(x)}) 
# multiply by 100 for percentage
asvTab.BAC.b.b.5 <- asvTab.BAC.b.b.4 %>% apply(2, function(x){x *100})
# classify as a data frame
asvTab.BAC.b.b.5 <- as.data.frame(asvTab.BAC.b.b.5)

# CTRL samples
########################################################
# calculate relative abundance of each cell by column value
asvTab.BAC.b.c.4 <- asvTab.BAC.b.c.3 %>% apply(2, function(x){x/sum(x)})
# multiply by 100 for percentage
asvTab.BAC.b.c.5 <- asvTab.BAC.b.c.4 %>% apply(2, function(x){x *100})
# classify as a data frame
asvTab.BAC.b.c.5 <- as.data.frame(asvTab.BAC.b.c.5)

# N samples
########################################################
# calculate relative abundance of each cell by column value
asvTab.BAC.b.n.4 <- asvTab.BAC.b.n.3 %>% apply(2, function(x){x/sum(x)})
# multiply by 100 for percentage
asvTab.BAC.b.n.5 <- asvTab.BAC.b.n.4 %>% apply(2, function(x){x *100})
# classify as a data frame
asvTab.BAC.b.n.5 <- as.data.frame(asvTab.BAC.b.n.5)

# P samples
########################################################
# calculate relative abundance of each cell by column value
asvTab.BAC.b.p.4 <- asvTab.BAC.b.p.3 %>% apply(2, function(x){x/sum(x)})
# multiply by 100 for percentage
asvTab.BAC.b.p.5 <- asvTab.BAC.b.p.4 %>% apply(2, function(x){x *100})
# classify as a data frame
asvTab.BAC.b.p.5 <- as.data.frame(asvTab.BAC.b.p.5)

# NP samples
########################################################
# calculate relative abundance of each cell by column value
asvTab.BAC.b.np.4 <- asvTab.BAC.b.np.3 %>% apply(2, function(x){x/sum(x)})
# multiply by 100 for percentage
asvTab.BAC.b.np.5 <- asvTab.BAC.b.np.4 %>% apply(2, function(x){x *100})
# classify as a data frame
asvTab.BAC.b.np.5 <- as.data.frame(asvTab.BAC.b.np.5)

```
  
```{r merge the taxonomy data frame w/ the ESV relative abundance (%) data frames}
# Bulk samples
########################################################
# move rownames of ASV table into 1st column as 'ASV_ID'
asvTab.BAC.b.b.5 <- tibble::rownames_to_column(asvTab.BAC.b.b.5, "ASV_ID")
# Left Join using inner_join function from dplyr
asvTab.BAC.b.b.5.T <- asvTab.BAC.b.b.5 %>% inner_join(taxTab, by= "ASV_ID")

# CTRL samples
########################################################
# move rownames of ASV table into 1st column as 'ASV_ID'
asvTab.BAC.b.c.5 <- tibble::rownames_to_column(asvTab.BAC.b.c.5, "ASV_ID")
# Left Join using inner_join function from dplyr
asvTab.BAC.b.c.5.T <- asvTab.BAC.b.c.5 %>% inner_join(taxTab, by= "ASV_ID")

# N samples
########################################################
# move rownames of ASV table into 1st column as 'ASV_ID'
asvTab.BAC.b.n.5 <- tibble::rownames_to_column(asvTab.BAC.b.n.5, "ASV_ID")
# Left Join using inner_join function from dplyr
asvTab.BAC.b.n.5.T <- asvTab.BAC.b.n.5 %>% inner_join(taxTab, by= "ASV_ID")

# P samples
########################################################
# move rownames of ASV table into 1st column as 'ASV_ID'
asvTab.BAC.b.p.5 <- tibble::rownames_to_column(asvTab.BAC.b.p.5, "ASV_ID")
# Left Join using inner_join function from dplyr
asvTab.BAC.b.p.5.T <- asvTab.BAC.b.p.5 %>% inner_join(taxTab, by= "ASV_ID")

# N samples
########################################################
# move rownames of ASV table into 1st column as 'ASV_ID'
asvTab.BAC.b.np.5 <- tibble::rownames_to_column(asvTab.BAC.b.np.5, "ASV_ID")
# Left Join using inner_join function from dplyr
asvTab.BAC.b.np.5.T <- asvTab.BAC.b.np.5 %>% inner_join(taxTab, by= "ASV_ID")

```
  
### Save files  
  
```{r save files}
# Bulk samples
########################################################
# save as an R file
saveRDS(asvTab.BAC.b.b.5.T, paste0(objects2.h.fp, "/asvTab.BAC.b.b.5.T.rds"))
#save as .txt file
write.table(asvTab.BAC.b.b.5.T, file = paste0(output2.h.fp , "/HoareBULK_16S_RelativeAbundance.csv"), 
            sep = ",", row.names = TRUE, col.names = NA)

# CTRL samples
########################################################
# save as an R file
saveRDS(asvTab.BAC.b.c.5.T, paste0(objects2.h.fp, "/asvTab.BAC.b.c.5.T.rds"))
#save as .txt file
write.table(asvTab.BAC.b.c.5.T, file = paste0(output2.h.fp , "/HoareCTRL_16S_RelativeAbundance.csv"), 
            sep = ",", row.names = TRUE, col.names = NA)

# N samples
########################################################
# save as an R file
saveRDS(asvTab.BAC.b.n.5.T, paste0(objects2.h.fp, "/asvTab.BAC.b.n.5.T.rds"))
#save as .txt file
write.table(asvTab.BAC.b.n.5.T, file = paste0(output2.h.fp , "/HoareN_16S_RelativeAbundance.csv"), 
            sep = ",", row.names = TRUE, col.names = NA)

# P samples
########################################################
# save as an R file
saveRDS(asvTab.BAC.b.p.5.T, paste0(objects2.h.fp, "/asvTab.BAC.b.p.5.T.rds"))
#save as .txt file
write.table(asvTab.BAC.b.p.5.T, file = paste0(output2.h.fp , "/HoareP_16S_RelativeAbundance.csv"), 
            sep = ",", row.names = TRUE, col.names = NA)

# NP samples
########################################################
# save as an R file
saveRDS(asvTab.BAC.b.np.5.T, paste0(objects2.h.fp, "/asvTab.BAC.b.np.6.T.rds"))
#save as .txt file
write.table(asvTab.BAC.b.np.5.T, file = paste0(output2.h.fp , "/HoareNP_16S_RelativeAbundance.csv"), 
            sep = ",", row.names = TRUE, col.names = NA)

```

## ESVs greater than 1% of the overall group sequences  
  
```{r filter for greater than one percent}
#bulk
asvTab.BAC.b.b.top <- asvTab.BAC.b.b.5.T %>% filter(total > 1.0)

#remove domain because all remaining are bacteria
asvTab.BAC.b.b.top <- asvTab.BAC.b.b.top %>% select(-Domain)

#control
asvTab.BAC.b.c.top <- asvTab.BAC.b.c.5.T %>% filter(total > 1.0)

#remove domain because all remaining are bacteria
asvTab.BAC.b.c.top <- asvTab.BAC.b.c.top %>% select(-Domain)

#N
asvTab.BAC.b.n.top <- asvTab.BAC.b.n.5.T %>% filter(total > 1.0)

#remove domain because all remaining are bacteria
asvTab.BAC.b.n.top <- asvTab.BAC.b.n.top %>% select(-Domain)

#P
asvTab.BAC.b.p.top <- asvTab.BAC.b.p.5.T %>% filter(total > 1.0)

#remove domain because all remaining are bacteria
asvTab.BAC.b.p.top <- asvTab.BAC.b.p.top %>% select(-Domain)

#NP
asvTab.BAC.b.np.top <- asvTab.BAC.b.np.5.T %>% filter(total > 1.0)

#remove domain because all remaining are bacteria
asvTab.BAC.b.np.top <- asvTab.BAC.b.np.top %>% select(-Domain)

```
  
```{r kable 1, include = TRUE}
#create table of library  
knitr::kable(asvTab.BAC.b.b.top, booktabs = T, digits = 2, caption = 'Most Relatively Abundant ASVs in Hoare basin bulk samples') %>%
  kable_styling(latex_options = c("striped", "scale_down", "hold_position"), font_size = 12) %>%
  row_spec(0, bold = T)

```
  
\newpage  
  
```{r kable 2, include = TRUE}
#create table of library  
knitr::kable(asvTab.BAC.b.c.top, booktabs = T, digits = 2, caption = 'Most Relatively Abundant ASVs in Hoare basin control samples') %>%
  kable_styling(latex_options = c("striped", "scale_down", "hold_position"), font_size = 12) %>%
  row_spec(0, bold = T)

```
  
\newpage  

```{r kable 3, include = TRUE}
#create table of library  
knitr::kable(asvTab.BAC.b.n.top, booktabs = T, digits = 2, caption = 'Most Relatively Abundant ASVs in Hoare basin N samples') %>%
  kable_styling(latex_options = c("striped", "scale_down", "hold_position"), font_size = 12) %>%
  row_spec(0, bold = T)

```

\newpage  
  
```{r kable 4, include = TRUE}
#create table of library  
knitr::kable(asvTab.BAC.b.p.top, booktabs = T, digits = 2, caption = 'Most Relatively Abundant ASVs in Hoare basin P samples') %>%
  kable_styling(latex_options = c("striped", "scale_down", "hold_position"), font_size = 12) %>%
  row_spec(0, bold = T)

```

\newpage  
   
```{r kable 5, include = TRUE}
#create table of library  
knitr::kable(asvTab.BAC.b.np.top, booktabs = T, digits = 2, caption = 'Most Relatively Abundant ASVs in Hoare basin NP samples') %>%
  kable_styling(latex_options = c("striped", "scale_down", "hold_position"), font_size = 12) %>%
  row_spec(0, bold = T)

```
  
```{r save top ASV files}
# Bulk samples
########################################################
# save as an R file
saveRDS(asvTab.BAC.b.b.top, paste0(objects2.h.fp, "/asvTab.BAC.b.b.top.rds"))
#save as .txt file
write.table(asvTab.BAC.b.b.top, file = paste0(output2.h.fp , "/HoareBULK_16S_Top.csv"), 
            sep = ",", row.names = TRUE, col.names = NA)

# CTRL samples
########################################################
# save as an R file
saveRDS(asvTab.BAC.b.c.top, paste0(objects2.h.fp, "/asvTab.BAC.b.c.top.rds"))
#save as .txt file
write.table(asvTab.BAC.b.c.top, file = paste0(output2.h.fp , "/HoareCTRL_16S_Top.csv"), 
            sep = ",", row.names = TRUE, col.names = NA)

# N samples
########################################################
# save as an R file
saveRDS(asvTab.BAC.b.n.top, paste0(objects2.h.fp, "/asvTab.BAC.b.n.top.rds"))
#save as .txt file
write.table(asvTab.BAC.b.n.top, file = paste0(output2.h.fp , "/HoareN_16S_Top.csv"), 
            sep = ",", row.names = TRUE, col.names = NA)

# P samples
########################################################
# save as an R file
saveRDS(asvTab.BAC.b.p.top, paste0(objects2.h.fp, "/asvTab.BAC.b.p.top.rds"))
#save as .txt file
write.table(asvTab.BAC.b.p.top, file = paste0(output2.h.fp , "/HoareP_16S_Top.csv"), 
            sep = ",", row.names = TRUE, col.names = NA)

# NP samples
########################################################
# save as an R file
saveRDS(asvTab.BAC.b.np.top, paste0(objects2.h.fp, "/asvTab.BAC.b.np.top.rds"))
#save as .txt file
write.table(asvTab.BAC.b.np.top, file = paste0(output2.h.fp , "/HoareNP_16S_Top.csv"), 
            sep = ",", row.names = TRUE, col.names = NA)

```
